Automattic/newspack-ads
# feat: header bidding

# Experimental header bidding support

This PR implements an experimental client-side header bidding support using [Prebid.js](https://prebid.org/) with the required modules for connecting with [Media.net](https://www.media.net/) as default bidder.

![image](https://user-images.githubusercontent.com/820752/144904676-4ca75bec-df4e-4984-8f9c-146187a28491.png)

## Implementation details

### Settings

Header bidding configuration settings are implemented using the `Newspack_Ads_Settings` feature. Additional bidder/partner configuration is also added using this system.

### Bidders/Partners

The bidder, or partner, is registered through the `newspack_register_ads_bidder()` function. [The default partner is implemented separately](https://github.com/Automattic/newspack-ads/pull/231/files#diff-c2ab2e6206485804b55dbfd303d2b072a7bd0f51cfa5747793332fbc78fea58f) for maintainability and also to provide an example of registering new bidders natively or through other plugins.

After registering through `newspack_register_ads_bidder()`, it should also hook into the [`newspack_ads_{$bidder_id}_ad_unit_bid`](https://github.com/Automattic/newspack-ads/pull/231/files#diff-46ecf0155737008dc15ef7553e83f14a15b501eb8b38f8d3cb326543947de987R185-R197) filter to apply the appropriate Prebid.js ad unit configuration. The default bidder also [provides a good example](https://github.com/Automattic/newspack-ads/pull/231/files#diff-c2ab2e6206485804b55dbfd303d2b072a7bd0f51cfa5747793332fbc78fea58fR74-R95) of how it works.

### Placements

The placements system has been extended to implement the bidder "placement ID". It's a unique ID provided by the bidder that should be used for identifying the exact bid with the provider.

Currently, each registered bidder generates an automatic `{$bidder_name} Placement ID` field inside a placement. It does not support extra bidder parameters per placement, as it has not yet been identified a reason to support such customization.

While selecting ad units, the UI detects if the registered bidder supports the selected ad unit sizes. If no size is supported, the bidder placement ID gets disabled and a warning is shown. It shouldn't affect the regular behavior of the ad unit and GPT, but also shouldn't attempt bidding.

![image](https://user-images.githubusercontent.com/820752/144490826-98ef9858-8962-4c15-b8d2-128ecf414f4c.png)

### GPT script

Hooks have been added to the GPT ad units and script configuration so that the header bidding implementation can append ad unit metadata and inject the Prebid.js script.

### Wizard UI

The placements UI has been updated to accommodate the increase of fields. The modal approach has been chosen to not flood the wizard screen with too many form fields and buttons.

![image](https://user-images.githubusercontent.com/820752/144487236-9a265a6a-b1c5-48c5-bc24-9c0dc3d6a1d1.png)

![image](https://user-images.githubusercontent.com/820752/144487277-a7c61982-6924-421e-a62a-0fe92b368764.png)

![image](https://user-images.githubusercontent.com/820752/144487359-6d551ce2-19b3-4cb5-9cff-662aadafed10.png)

## Experimental

This is an experimental feature but should still be able to function while disabled. For that reason, the method we'll temporarily use for disabling it is to consider that there are no registered bidders available.

To enable the bidder registration you must add this to your `wp-config.php`:

```php
define( 'NEWSPACK_ADS_EXPERIMENTAL_MEDIANET', true );
```

## Testing

The first test is to make sure the changes here don't affect ads configuration or display, other than the placements UI updates.

**Before adding the `NEWSPACK_ADS_EXPERIMENTAL_MEDIANET` constant**:

1. Check out this branch and https://github.com/Automattic/newspack-plugin/pull/1285
2. Visit **Advertising -> Settings**
3. Ensure no Header Bidding section is visible
4. Visit **Advertising -> Placements**
5. Confirm you see the new UI as shown on the images above
6. Click on the "Placement settings" button of a placement, confirm no "Media.net" field appears
7. Select an ad unit and save
8. Confirm that the placement works as expected with the selected ad unit

### Header bidding settings

1. Define `NEWSPACK_ADS_EXPERIMENTAL_MEDIANET` as described above
2. Visit **Advertising -> Settings**
3. Enable **Header Bidding** and do not change any setting just yet
4. Visit **Advertising -> Placements**, edit a placement, and confirm there's still **no Media.net field**. It should still require the **Media.net Customer ID**.
5. Back to **Advertising -> Settings**, fill the **Media.net Customer ID** with a random `1234567890`, enable **Debug mode**, and click **Save settings**
6. Visit **Advertising -> Placements**, edit a valid placement and confirm you see the **Media.net Placement ID** field


### Placement editing and debug mode

1. Still editing the placement, select different ad units
2. Confirm you see the warning message when ad units with non-standard sizes are selected
3. Save the placement with a valid ad unit and a random `1234567890` as the **Media.net Placement ID**
4. Visit a page where this placement renders the ad, open `DevTools` and confirm there are no errors and you see debug logs similar to this:

![image](https://user-images.githubusercontent.com/820752/144493713-a42b0970-7181-4fd8-be2e-c36b970192ce.png)

No actual bids should take place, as we are using fake IDs.

### Unit tests

[Written unit tests](https://github.com/Automattic/newspack-ads/pull/231/files#diff-f69462aa64c5b2a82c9734232ea41161a6a43231ac6ec38ac9c67a0a877a931e) should also pass.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201230241510067
